File: algorithm/Theme/FenwickTree/fenwick_tree.pl
#!/usr/bin/env perl -n
#===============================================================================
#         FILE: fenwick_tree.pl
#       AUTHOR: Sabastian (liuxueyang.github.io), liuxueyang457@gmail.com
# ORGANIZATION: Hunan University
#      CREATED: 2016/06/26 23时27分06秒
#===============================================================================

use strict;
use warnings;
use utf8;
use 5.014;

chomp;
my @fenwick;
my @array = split;
unshift @array, 0;
unshift @fenwick, 0;

sub lowbit { $_[0] & -$_[0]; }

sub build_fenwick_tree {
    for (1..$#array) {
        for my $j ($_-lowbit($_)+1..$_) {
            $fenwick[$_] += $array[$j]
        }
    }
}

sub modify_fenwick_tree {
    # two arguments
    # 1.index: the index of the array to modify
    # 2.delta: substract result of new value to the old value of the element
    my ($index, $delta) = @_;
    while ($index <= @fenwick) {
        $fenwick[$index] += $delta;
        $index += lowbit($index);
    }
}

sub sum_fenwick_tree {
    # 1 argument: index
    my ($ans, $i) = (0, $_[0]);
    while ($i > 0) {
        $ans += $fenwick[$i];
        $i -= lowbit($i);
    }
    $ans;
}


build_fenwick_tree();
print join(" ", @fenwick[1..$#fenwick]), "\n";

modify_fenwick_tree(3, 3);
print join(" ", @fenwick[1..$#fenwick]), "\n";

print sum_fenwick_tree(3), "\n";

# input
# 1 2 3 4 5
# output
# 1 3 3 10 5
# 1 3 6 13 5
# 9
#

File: algorithm/Theme/FenwickTree/fenwick_tree.c
#include <stdio.h>

const int N = 6;
int a[N] = {0, 1, 2, 3, 4, 5};
int fenwick[N];

int lowbit(int x) {
  return x & (-x);
}

void build_fenwick_tree() {
    for (int i = 1; i < N; ++i) {
      fenwick[i] = a[i];
      for (int j = i - 1; j > i - lowbit(i); --j) {
        fenwick[i] += a[j];
      }
  }
}

void update_fenwick_tree(int index, int delta) {
  while (index < N) {
    fenwick[index] += delta;
    index += lowbit(index);
  }
}

int sum_fenwick_tree(int index) {
  int ans = 0;
  while (index > 0) {
    ans += fenwick[index];
    index -= lowbit(index);
  }
  return ans;
}

void print_fenwick_tree() {
  for (int i = 1; i < N; ++i) {
    printf("%d ", fenwick[i]);
  }
  printf("\n");
}

int main(void) {
  build_fenwick_tree();
  print_fenwick_tree();

  update_fenwick_tree(3, 3);
  print_fenwick_tree();

  printf("%d\n", sum_fenwick_tree(4));
  printf("%d\n", sum_fenwick_tree(5));

  return 0;
}
